### PR TITLE
ICMSLST-444 Add/remove contacts for exporter

### PR DESCRIPTION
**Pull request by kevinetienne**

### CODE REVIEW

File: web/domains/exporter/forms.py [hunk 1]
```diff
@@ -15,12 +19,44 @@ class Meta:
...
+        "department__icontains",
+    ]
+
+    def label_from_instance(self, user):
+        return f"{user.full_name} - {user.email}"
+
+
+class ExporterForm(ModelForm):
+    contacts_to_add = ModelMultipleChoiceField(
+        queryset=User.objects.all(), widget=ContactWidget, required=False
```
Comment (kevinetienne) [inline web/domains/exporter/forms.py:38]: Filter by account status

Comment (kevinetienne) [reply web/domains/exporter/forms.py:38]: Plus exclude people already assigned as contacts

File: web/domains/exporter/forms.py [hunk 2]
```diff
@@ -15,12 +19,44 @@ class Meta:
...
+        "first_name__icontains",
+        "last_name__icontains",
+        "email__icontains",
+        "job_title__icontains",
+        "organisation__icontains",
+        "department__icontains",
+    ]
+
+    def label_from_instance(self, user):
+        return f"{user.full_name} - {user.email}"
```
Comment (kevinetienne) [inline web/domains/exporter/forms.py:33]: move to `widgets.py`

File: web/domains/exporter/urls.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 urlpatterns = [
     path("", views.ExporterListView.as_view(), name="exporter-list"),
-    path("<int:pk>/edit/", views.ExporterEditView.as_view(), name="exporter-edit"),
-    path("new/", views.ExporterCreateView.as_view(), name="exporter-new"),
+    path("<int:pk>/edit/", views.edit, name="exporter-edit"),
+    path("create/", views.create, name="exporter-create"),
```
Comment (kevinetienne) [inline web/domains/exporter/urls.py:11]: 2 new endpoints:
 - add contact for exporter
 - delete contact for exporter

File: web/domains/exporter/views.py [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
-from web.views import ModelCreateView, ModelDetailView, ModelFilterView, ModelUpdateView
+from web.views import ModelDetailView, ModelFilterView
 from web.views.actions import Archive, Edit, Unarchive
 
-from .forms import ExporterEditForm, ExporterFilter
+from .forms import ExporterForm, ExporterFilter
 from .models import Exporter
 
-permissions = "web.IMP_MAINTAIN_ALL"
+permissions = "web.exporter_access"
```
Comment (kevinetienne) [inline web/domains/exporter/views.py:13]: use inline

File: web/domains/exporter/views.py [hunk 5]
```diff
@@ -22,28 +26,47 @@ class Display:
...
-class ExporterEditView(ModelUpdateView):
-    template_name = "web/domains/exporter/edit.html"
-    form_class = ExporterEditForm
-    success_url = reverse_lazy("exporter-list")
-    cancel_url = success_url
-    model = Exporter
-    permission_required = permissions
+@login_required
+@permission_required("web.exporter_access", raise_exception=True)
+def edit(request, pk):
```
Comment (kevinetienne) [inline web/domains/exporter/views.py:31]: `edit_exporter`

File: web/domains/exporter/views.py [hunk 6]
```diff
@@ -22,28 +26,47 @@ class Display:
...
-    form_class = ExporterEditForm
-    success_url = reverse_lazy("exporter-list")
-    cancel_url = success_url
-    model = Exporter
-    permission_required = permissions
-    page_title = "Create Exporter"
+
+@login_required
+@permission_required("web.exporter_access", raise_exception=True)
+def create(request):
```
Comment (kevinetienne) [inline web/domains/exporter/views.py:53]: `create_exporter`

File: web/domains/exporter/views.py [hunk 7]
```diff
@@ -22,28 +26,47 @@ class Display:
...
+    form = ExporterForm()
+    if request.POST:
+        form = ExporterForm(request.POST)
+        if form.is_valid():
+            form.save()
+            return redirect(reverse("exporter-list"))
```
Comment (kevinetienne) [inline web/domains/exporter/views.py:59]: go to edit

File: web/domains/exporter/views.py [hunk 8]
```diff
@@ -22,28 +26,47 @@ class Display:
...
+    return render(request, "web/domains/exporter/create.html", context)
 
 
 class ExporterDetailView(ModelDetailView):
     template_name = "web/domains/exporter/view.html"
-    form_class = ExporterEditForm
+    form_class = ExporterForm
     cancel_url = reverse_lazy("exporter-list")
     model = Exporter
     permission_required = permissions
```
Comment (kevinetienne) [inline web/domains/exporter/views.py:72]: remove

File: web/templates/partial/exporter/sidebar.html [hunk 9]
```diff
@@ -0,0 +1,15 @@
+{% block sidebar %}
+  <h4>Exporter</h4>
+  <h5>Search</h5>
+  <ul class="menu-out">
+    <li>
+        <a href="{{ url('exporter-list') }}">Access Requests</a>
```
Comment (kevinetienne) [inline web/templates/partial/exporter/sidebar.html:6]: It's not Access Requests

File: web/templates/web/contacts/grid.html [hunk 10]
```diff
@@ -1,33 +1,21 @@
 <div id="team-members" class="tab-content big-tabs" aria-hidden="false">
-    {% if not members %}
+    {% if not contacts %}
     <div class="info-box info-box-info">
         There isn't anyone in this team
```
Comment (kevinetienne) [inline web/templates/web/contacts/grid.html:4]: There is no contact attached to this exporter

File: web/templates/web/contacts/grid.html [hunk 11]
```diff
@@ -1,33 +1,21 @@
...
-                        <div class="input-group-addon hint-addon">
-                            <div id="hint{{role.id}}-content" role="tooltip" class="hint-content">{{role.description}}</div>
-                            <div id="hint{{role.id}}" aria-describedby="hint{{role.id}}-content" class="hint icon-info"></div>
-                        </div>
-                    </div>
-                </th>
-                {% endfor %}
+              <th>Central Contact Details</th>
+              {% if not read_only %}
+                <th>Delete contact</th>
```
Comment (kevinetienne) [inline web/templates/web/contacts/grid.html:12]: Actions

File: web/tests/domains/exporter/test_views.py [hunk 12]
```diff
@@ -1,9 +1,11 @@
+from guardian.shortcuts import assign_perm
+
 from web.domains.exporter.models import Exporter
 from web.tests.auth import AuthTestCase
 from web.tests.domains.exporter.factory import ExporterFactory
 
 LOGIN_URL = "/"
-PERMISSIONS = ["IMP_MAINTAIN_ALL"]
+PERMISSIONS = ["exporter_access"]
```
Comment (kevinetienne) [inline web/tests/domains/exporter/test_views.py:8]: inline

File: web/domains/exporter/views.py [hunk 13]
```diff
@@ -20,28 +26,72 @@ class Display:
...
-    model = Exporter
-    permission_required = "web.reference_data_access"
-    page_title = "Create Exporter"
+    exporter_contacts = get_users_with_perms(
+        exporter, only_with_perms_in=["is_contact_of_exporter"]
+    ).filter(user_permissions__codename="exporter_access")
+    available_contacts = (
+        User.objects.account_active()
+        .filter(user_permissions__codename="exporter_access")
+        .exclude(pk__in=exporter_contacts)
```
Comment (oskusalerma) [inline web/domains/exporter/views.py:47]: can i see the sql for this just to verify it's doing what we think it's doing

File: web/domains/exporter/views.py [hunk 14]
```diff
@@ -20,28 +26,72 @@ class Display:
...
-    template_name = "web/domains/exporter/view.html"
-    form_class = ExporterEditForm
-    cancel_url = reverse_lazy("exporter-list")
-    model = Exporter
-    permission_required = "web.reference_data_access"
+
+@login_required
+@permission_required("web.reference_data_access", raise_exception=True)
+def create_exporter(request):
+    form = ExporterForm()
```
Comment (oskusalerma) [inline web/domains/exporter/views.py:62]: same thing here, put in else

File: web/domains/user/managers.py [hunk 15]
```diff
@@ -31,3 +31,6 @@ def create_superuser(self, username, password=None, **extra_fields):
         if extra_fields.get("is_superuser") is not True:
             raise ValueError("Superuser must have is_superuser=True.")
         return self.create_user(username, password, **extra_fields)
+
+    def account_active(self):
+        return self.filter(account_status="ACTIVE")
```
Comment (oskusalerma) [inline web/domains/user/managers.py:36]: try doing self.model.ACTIVE?

File: web/templates/partial/contacts/contacts.html [hunk 16]
```diff
@@ -1,3 +1,85 @@
 <h3>Contacts</h3>
 
-{% include "web/contacts/tabs.html" %}
+<div id="team-members" class="tab-content big-tabs" aria-hidden="false">
+    {% if not contacts %}
+    <div class="info-box info-box-info">
+        There are no contacts attached to this exporter
```
Comment (oskusalerma) [inline web/templates/partial/contacts/contacts.html:6]: reword this so we can share with importer

File: web/templates/partial/contacts/contacts.html [hunk 17]
```diff
@@ -1,3 +1,85 @@
...
+                        </div>
+                    </div>
+
+                </td>
+                {% if not read_only %}
+                  <td>
+                    <div class="row">
+                        <div class="twelve columns">
+                          <form method="post"
+                                action="{{ url('exporter-contact-delete', kwargs={'pk': object.pk, 'contact_pk': user.pk}) }}">
```
Comment (oskusalerma) [inline web/templates/partial/contacts/contacts.html:58]: on the import ticket you might have to move this out to a block so importer can have its own urls

**Review by oskusalerma — CHANGES_REQUESTED** [review]

File: web/domains/exporter/views.py (hunk 13 above)
Comment (kevinetienne) [reply web/domains/exporter/views.py:47]: This is the query

```python
from guardian.shortcuts import get_users_with_perms
from web.models import User


>> qs = get_users_with_perms(Exporter.objects.get(pk=16), only_with_perms_in=["is_contact_of_exporter"]).filter(user_permissions__codename="exporter_access")

>> User.objects.account_active().filter(user_permissions__codename="exporter_access").exclude(pk__in=qs)
<QuerySet []>

>> str(User.objects.account_active().filter(user_permissions__codename="exporter_access").exclude(pk__in=qs).query)
```

```sql
'SELECT * 
  FROM "web_user"
  INNER JOIN "web_user_user_permissions" ON ("web_user"."id" = "web_user_user_permissions"."user_id")
  INNER JOIN "auth_permission" ON ("web_user_user_permissions"."permission_id" = "auth_permission"."id") 
  WHERE ("web_user"."account_status" = ACTIVE AND "auth_permission"."codename" = exporter_access AND NOT ("web_user"."id" IN (SELECT DISTINCT V0."id" FROM "web_user" V0 LEFT OUTER JOIN "guardian_userobjectpermission" V1 ON (V0."id" = V1."user_id") LEFT OUTER JOIN "web_user_groups" V4 ON (V0."id" = V4."user_id") INNER JOIN "web_user_user_permissions" V6 ON (V0."id" = V6."user_id") INNER JOIN "auth_permission" V7 ON (V6."permission_id" = V7."id") WHERE ((V1."content_type_id" = 33 AND V1."object_pk" = 16 AND V1."permission_id" IN (SELECT U0."id" FROM "auth_permission" U0 WHERE (U0."codename" IN (is_contact_of_exporter) AND U0."content_type_id" = 33))) AND V7."codename" = exporter_access)))) ORDER BY "web_user"."is_active" DESC, "web_user"."first_name" ASC'
```

Not that:
```python
>> qs = get_users_with_perms(Exporter.objects.get(pk=16), only_with_perms_in=["is_contact_of_exporter"]).filter(user_permissions__codename="exporter_access")
```
generates 

```sql
'SELECT * FROM "web_user" INNER JOIN "web_user_user_permissions" ON ("web_user"."id" = "web_user_user_permissions"."user_id") INNER JOIN "auth_permission" ON ("web_user_user_permissions"."permission_id" = "auth_permission"."id") WHERE ("web_user"."account_status" = ACTIVE AND "auth_permission"."codename" = exporter_access AND NOT ("web_user"."id" IN (SELECT DISTINCT V0."id" FROM "web_user" V0 LEFT OUTER JOIN "guardian_userobjectpermission" V1 ON (V0."id" = V1."user_id") LEFT OUTER JOIN "web_user_groups" V4 ON (V0."id" = V4."user_id") INNER JOIN "web_user_user_permissions" V6 ON (V0."id" = V6."user_id") INNER JOIN "auth_permission" V7 ON (V6."permission_id" = V7."id") WHERE ((V1."content_type_id" = 33 AND V1."object_pk" = 16 AND V1."permission_id" IN (SELECT U0."id" FROM "auth_permission" U0 WHERE (U0."codename" IN (is_contact_of_exporter) AND U0."content_type_id" = 33))) AND V7."codename" = exporter_access)))) ORDER BY "web_user"."is_active" DESC, "web_user"."first_name" ASC'
```

which is what you see partially after the `NOT` in the where clause.

File: web/templates/partial/contacts/contacts.html (hunk 16 above)
Comment (kevinetienne) [reply web/templates/partial/contacts/contacts.html]: Fixed in https://github.com/uktrade/icms/pull/196/commits/6226c0bd9fc2d3bf6e5ce36749827bb0bd6dc041

File: web/domains/user/managers.py (hunk 15 above)
Comment (kevinetienne) [reply web/domains/user/managers.py]: Fixed in https://github.com/uktrade/icms/pull/196/commits/6226c0bd9fc2d3bf6e5ce36749827bb0bd6dc041

File: web/domains/exporter/views.py (hunk 14 above)
Comment (kevinetienne) [reply web/domains/exporter/views.py]: Updated in https://github.com/uktrade/icms/pull/196/commits/6226c0bd9fc2d3bf6e5ce36749827bb0bd6dc041

**Review by oskusalerma — APPROVED** [review]

**Pull request merged**